File: problems/04-prefix-sum/001-subarray-sum-equals-k/java/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> frequency = new HashMap<>();
        frequency.put(0, 1);

        int prefix = 0;
        int count = 0;

        for (int value : nums) {
            prefix += value;

            int needed = prefix - k;
            count += frequency.getOrDefault(needed, 0);

            frequency.put(prefix, frequency.getOrDefault(prefix, 0) + 1);
        }

        return count;
    }
}

File: problems/04-prefix-sum/001-subarray-sum-equals-k/python/solution.py
class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        frequency = {0: 1}
        prefix = 0
        count = 0

        for value in nums:
            prefix += value

            needed = prefix - k
            count += frequency.get(needed, 0)

            frequency[prefix] = frequency.get(prefix, 0) + 1

        return count
